File: backend/models/restaurant.py
from pydantic import BaseModel, Field, HttpUrl, ConfigDict
from typing import Any, Optional
from bson import ObjectId

class Position(BaseModel):
    """Position model with latitude and longitude coordinates"""
    lat: float
    lng: float

class RestaurantBase(BaseModel):
    """Base model with common restaurant attributes"""
    name: str
    img_url: HttpUrl = Field(..., description="URL to restaurant image")
    rating: Optional[float] = None
    type: Optional[str] = None
    description: Optional[str] = None
    position: Optional[Position] = None

class RestaurantCreate(RestaurantBase):
    """Model used for restaurant creation requests"""
    class Config:
        json_encoders = {
            ObjectId: lambda v: str(v),  # Convert ObjectId to string for JSON serialization
            HttpUrl: lambda v: str(v)  # Convert HttpUrl to string for JSON serialization
        }
    pass

class RestaurantInDB(RestaurantBase):
    """Model representing how restaurant is stored in database"""
    id: str = Field(default=None, alias="_id")
    
    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True
    )

class Restaurant(RestaurantBase):
    """Model returned to clients via API"""
    id: str

    model_config = ConfigDict(populate_by_name=True)
        
    @classmethod
    def from_db_model(cls, db_restaurant: RestaurantInDB) -> "Restaurant":
        """Convert from database model to API model"""
        return cls(
            id=str(db_restaurant.id),
            name=db_restaurant.name,
            img_url=db_restaurant.img_url,
            rating=db_restaurant.rating,
            type=db_restaurant.type,
            description=db_restaurant.description,
            position=db_restaurant.position
        )

File: ratingCronJob/pkg/model/restaurant.go
package model

import "go.mongodb.org/mongo-driver/v2/bson"

// Position represents geographic coordinates
type Position struct {
	Lat float64 `bson:"lat" json:"lat"`
	Lng float64 `bson:"lng" json:"lng"`
}

// Restaurant represents a restaurant entity
type Restaurant struct {
	ID          bson.ObjectID	   `bson:"_id,omitempty" json:"id,omitempty"`
	Name        string             `bson:"name" json:"name"`
	ImgURL      string             `bson:"img_url" json:"img_url"`
	Rating      float64            `bson:"rating" json:"rating"`
	Type        string             `bson:"type,omitempty" json:"type,omitempty"`
	Description string             `bson:"description,omitempty" json:"description,omitempty"`
	Position    *Position          `bson:"position,omitempty" json:"position,omitempty"`
	ReviewCount int                `bson:"_review_count" json:"_review_count"`
}

